File: cpp/src/fps.cpp
#include "fps.h"

FPS::FPS() {

    this->frame_count = 0;

    this->fps = 0.0f;
    this->avg_fps = 0.0f;
    this->frame_time = 0.0f;
}

FPS::~FPS() {}

void FPS::start() {

    this->start_time = std::chrono::system_clock::now();

    return;
}

void FPS::update() {

    this->current_time = std::chrono::system_clock::now();
    this->elapsed_time = this->current_time - this->start_time;

    this->frame_time += this->elapsed_time.count();

    this->frame_count++;

    return;
}

float FPS::getFps() {

    if(this->frame_time > 0.0f){
        this->fps = this->frame_count / this->frame_time;
    }

    return this->fps;
}

float FPS::getAvgFps() {

    if(this->frame_count >= 30){
        this->avg_fps = this->frame_count / this->frame_time;
        this->frame_time = 0.0f;
        this->frame_count = 0;
    }

    return this->avg_fps;
}

void FPS::reset() {

    this->frame_count = 0;

    this->fps = 0.0f;
    this->avg_fps = 0.0f;
    this->frame_time = 0.0f;

    return ;
}

File: python/modules/fps.py
import datetime

from abc import ABC

class FPS(ABC):

    def __init__(self) -> None:

        self.start_time : datetime.datetime = None
        self.current_time : datetime.datetime = None
        self.elapsed_time : datetime.timedelta = None

        self.frame_count : int = 0

        self.fps : float = 0.0
        self.avg_fps : float = 0.0
        self.frame_time : float = 0.0
    
    def start(self) -> None:

        self.start_time = datetime.datetime.now()

        return 
    
    def update(self) -> None:

        self.current_time = datetime.datetime.now()
        self.elapsed_time = self.current_time - self.start_time

        self.frame_time += self.elapsed_time.total_seconds()

        self.frame_count += 1

        return 
    
    def get_fps(self) -> float:

        if self.frame_time > 0.0:
            self.fps = self.frame_count / self.frame_time

        return round(self.fps, 2)
    
    def get_avg_fps(self) -> float:

        if self.frame_count >= 30:

            self.avg_fps = self.frame_count / self.frame_time
            self.frame_time = 0.0
            self.frame_count = 0

        return round(self.avg_fps, 2)

    def reset(self) -> None:

        self.start_time = None
        self.current_time = None
        self.elapsed_time = None

        self.frame_count = 0

        self.fps = 0.0
        self.avg_fps = 0.0
        self.frame_time = 0.0

        return 